# 2025 Week-1 Homework

## Q1. Pandas version

In [24]:
import pandas as pd
import numpy as np

In [25]:
pd.__version__

'2.3.1'

## Getting the Car Fuel Efficiency dataset from the downloaded csv file

In [26]:
df = pd.read_csv("01-Intro/car_fuel_efficiency.csv")
df.head() # confirm data is read into the dataset

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

## Q2. Records count - How many records are in the dataset?

In [27]:
len(df)

9704

## Q3. - Fuel types: How many fuel types are presented in the dataset?

In [28]:
df["fuel_type"].nunique()

2

## Q4. Missing values: How many columns in the dataset have missing values?

In [29]:
nulls_per_series = df.isnull().sum()
series_with_nulls = (nulls_per_series > 0).sum()
print(series_with_nulls) 

4


## Q5. Max fuel efficiency: What's the maximum fuel efficiency of cars from Asia?

In [30]:
Asian_cars_df = df[df['origin'] == 'Asia'] # All Asian cars
max_fuel_efficiency = Asian_cars_df['fuel_efficiency_mpg'].max() # Maximum fuel efficiency of all Asian cars

print((max_fuel_efficiency))

23.759122836520497


## Q6. Median value of horsepower

#### 1. Find the median value of horsepower column in the dataset.

In [31]:
print(df.horsepower.median())

149.0


#### 2. Next, calculate the most frequent value of the same horsepower column

In [32]:
freq_hp_df = df['horsepower'].mode()[0]
print(freq_hp_df)

152.0


#### 3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.

In [33]:
df['horsepower'] = df['horsepower'].fillna(freq_hp_df)

#### 4. Now, calculate the median value of horsepower once again.

In [34]:
print(df.horsepower.median())

152.0


#### Yes, the median has increased to 152 from the previous value of 149

### Q7. Sum of weights

#### 1. Select all the cars from Asia

In [47]:
asia_cars_df = df[df['origin'] == 'Asia']
    
# [['vehicle_weight','model_year']].head(7)
asia_cars_df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       152.0     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

#### 2. Select only columns vehicle_weight and model_year

In [36]:
asian_cars_wght_year_df = asia_cars_df[['vehicle_weight','model_year']]
asian_cars_wght_year_df

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

#### 3. Select the first 7 values

In [37]:
asian_cars_top_7_df = asian_cars_wght_year_df.head(7)
asian_cars_top_7_df

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

#### 4. Get the underlying NumPy array. Let's call it X

In [38]:
X = asian_cars_top_7_df.round(2).to_numpy()
X

array([[2714.22, 2016.  ],
       [2783.87, 2010.  ],
       [3582.69, 2007.  ],
       [2231.81, 2011.  ],
       [2659.43, 2016.  ],
       [2844.23, 2014.  ],
       [3761.99, 2019.  ]])

#### 5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [48]:
XT = X.T # Let us get the X Transpose, first
XT

array([[2714.22, 2783.87, 3582.69, 2231.81, 2659.43, 2844.23, 3761.99],
       [2016.  , 2010.  , 2007.  , 2011.  , 2016.  , 2014.  , 2019.  ]])

In [49]:
XTX = XT.dot(X) # The matrix-matrix multiplication
XTX

array([[62248346.87540001, 41431222.87      ],
       [41431222.87      , 28373339.        ]])

#### 6. Invert XTX

In [50]:
XTX_inverse = np.linalg.inv(XTX)
XTX_inverse

array([[ 5.71499054e-07, -8.34512450e-07],
       [-8.34512450e-07,  1.25381335e-06]])

#### 7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].

In [52]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

#### 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w

In [57]:
w = XTX_inverse.dot(XT)
w

array([[-1.31202939e-04, -8.63909553e-05,  3.72637456e-04,
        -4.02727235e-04, -1.62515372e-04, -5.52333222e-05,
         4.65093087e-04],
       [ 2.62637337e-04,  1.96990665e-04, -4.73396011e-04,
         6.58945422e-04,  3.08360275e-04,  1.51634747e-04,
        -6.07978333e-04]])

In [58]:
result = w.dot(y)
result

array([0.01386366, 0.5049075 ])

#### 9. What's the sum of all the elements of the result?

In [65]:
result.sum()

np.float64(0.5187711601900705)